### ISSUE TITLE
[tail plugin] feature: group by regex

### ISSUE THREAD

**Issue by zypA13510** [collectd/collectd#3300]
Not a bug so please forgive me for not using the template.

Same idea as [pretaweb/pretaweb.collectd.groupingtail](/pretaweb/pretaweb.collectd.groupingtail)(this is not licensed, also doesn't seem to be maintained), instead of reporting on a single metric, I need the metrics grouped by another field in the log.

Example use cases:
- For a web server, monitor response size based on request path/user-agent string/source IP (what is common: they cannot be predefined and can only be extracted from the log). This doesn't seem possible currently.

**Comment by kwiatrox** [comment]
Please take a look at logparser plugin #3045 . Maybe it could be useful for your case.
It can parse values from log using regexps and group them by fields in the log (using instance or type), but provides notifications not metrics.

**Comment by zypA13510** [comment]
> but provides notifications not metrics
The rest of it sounds like exactly what I need, but I need metrics, not notifications.

**Comment by rpv-tomsk** [comment]
>what is common: they cannot be predefined and can only be extracted from the log

What you want will result in an uncontrollable number of metrics, at least for "request path/user-agent string". 

Source IP can be predefined.

To get controllable results you need to predefine groups.

**Comment by zypA13510** [comment]
> What you want will result in an uncontrollable number of metrics

Yes, that's true. Is there no way to generate an uncontrollable number of metrics, put in another word, dynamic metric creation in collectd?

> Source IP can be predefined.

How? If your server listen on `0.0.0.0` (or `::` for v6), how can you predefine billions of possible address? Although in practice, it's for a few authorized users only, and thus the pool will be limited to < 100, I don't think having to change the configuration just because one user changed his IP is a flexible and maintainable solution. Also it creates a lot of redundancy in the config.



Btw, maybe UA string is a bad choice of example, it'd be too diverse to be meaningful.

**Comment by rpv-tomsk** [comment]
>How? If your server listen on 0.0.0.0 (or :: for v6), how can you predefine billions of possible address?

To continue discussion, please explain me what do you want to do with billions of possible metrics, based on billions possible addresses?

**Comment by zypA13510** [comment]
As I said, there are only a few authorized users, their IP doesn't change often, but does change slowly over time. Basically, what you're saying is add all of them manually and maintain them manually, but why all the work if all of them are in the log file already? Why can't we just categorize using the log file? If I need to change my configuration every time a user changes their IP address, things become hard to maintain very fast. Same with the request path, even if I'm willing to do double the work when adding a new route/endpoint/whatever to the server, that's just unreliable as heck.

I can do this with `fluentd` just fine, but unfortunately, GCP only accepts metric recording from `collectd`, and log ingestion via `fluentd`.

Feel free to close this if you're still not very convinced and you don't like feature wishlist in issues. I understand this is open-source but I don't know much about C, can't help on the implementation.

**Comment by rpv-tomsk** [comment]
1)

>what is common: they cannot be predefined and can only be extracted from the log

2)
>how can you predefine billions of possible address

3)
>Btw, maybe UA string is a bad choice of example, it'd be too diverse to be meaningful.

You asked strange things and get strange answers then.

If you started discussion from your real use-case, we would not have wasted the time on these.

>Although in practice, it's for a few authorized users only, and thus the pool will be limited to < 100, I don't think having to change the configuration just because one user changed his IP is a flexible and maintainable solution. Also it creates a lot of redundancy in the config.

You should have started with just that.
With such update it makes sence, but currently not implemented in any form.
